Added <event_identifier> with <event_id>

enigmaeth opened the pull request:
<!-- 
(Thanks for sending a pull request! Please make sure you click the link above to view the contribution guidelines, then fill out the blanks below.)
-->

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `nextgen` branch.
- [x] The unit tests pass locally with my changes <!-- use `nosetests tests/unittests` to run all the tests -->

#### Short description of what this resolves:
Allowed using `event_identifier` wherever `event_id` is used

<!-- Add the issue number that is fixed by this PR (In the form Fixes #123) -->

Fixes #3808.
Fixes #3712.

poush commented on app/api/__init__.py:
```diff
@@ -102,19 +102,20 @@
 api.route(MicrolocationList, 'microlocation_list', '/microlocations', '/sessions/<int:id>/microlocations')
 api.route(MicrolocationDetail, 'microlocation_detail', '/microlocations/<int:id>',
-          '/events/<int:event_id>/microlocations', '/events/<identifier>/microlocations',
+          '/events/<int:event_id>/microlocations', '/events/<event_identifier>/microlocations',
```
It says `microlocations` while it is associated with MicroLocationDetail. Can you please check?

poush commented on app/api/__init__.py:
```diff
@@ -102,19 +102,20 @@
 
 # sessions
-api.route(SessionList, 'session_list', '/sessions', '/events/<int:event_id>/sessions', '/events/<identifier>/sessions',
+api.route(SessionList, 'session_list', '/sessions', '/events/<int:id>/sessions', '/events/<identifier>/sessions',
```
Please don't change it back to `id` as we already have to correct this to `event_id` in separate PR as well.

enigmaeth commented:
@fossasia/open-event-server-devs What are your ideas on how to add `event-identifier` to the apib docs?
 We need to provide it as an alternate option to a required field `event_id`.

shubham-padia commented:
Adding it as a seperate endpoint right now would be the way for now.

niranjan94 commented:
@enigmaeth in docs, right now we have 

```md
## Event Details [/v1/events/{event_id}]
+ Parameters
    + event_id: 1 (integer) - ID of the event in the form of an integer
```

Instead we could have 

```md
## Event Details [/v1/events/{event_identifier}]
+ Parameters
    + event_identifier: a31dda (string) - identifier of the event. (Can also be the event ID)
```

enigmaeth commented:
@niranjan94 Updated documentation. Please check.

niranjan94 commented on app/api/__init__.py:
```diff
@@ -47,7 +47,7 @@
 api.route(AllTicketList, 'all_ticket_list', '/tickets', '/events/<int:id>/tickets', '/events/<identifier>/tickets')
 api.route(TicketDetail, 'ticket_detail', '/tickets/<int:id>', '/events/<int:event_id>/tickets',
-          '/attendees/<int:attendee_id>/ticket')
+          '/events/<event_identifier>/tickets', '/attendees/<int:attendee_id>/ticket')
```
This is ticket detail. Why is `/events/<event_identifier>/tickets` needed here when it's already there for TicketList ?

niranjan94 commented on app/api/__init__.py:
```diff
@@ -97,19 +97,19 @@
 api.route(MicrolocationDetail, 'microlocation_detail', '/microlocations/<int:id>',
-          '/events/<int:event_id>/microlocations', '/events/<identifier>/microlocations',
-          '/sessions/<int:session_id>/microlocation')
+          '/events/<int:event_id>/microlocations', '/sessions/<int:session_id>/microlocation')
```
Why is `/events/<int:event_id>/microlocations` here ? This is microlocation detail. This should be in list.

niranjan94 commented on app/api/__init__.py:
```diff
@@ -97,19 +97,19 @@
           '/microlocations/<int:microlocation_id>/sessions')
 api.route(SessionDetail, 'session_detail', '/sessions/<int:id>',
-          '/microlocations/<int:microlocation_id>/sessions', '/events/<int:event_id>/microlocations')
+          '/microlocations/<int:microlocation_id>/sessions', '/events/<int:event_id>/microlocations',
```
Why are `/microlocations/<int:microlocation_id>/sessions`, `/events/<int:event_id>/microlocations` & `/events/<event_identifier>/microlocations` even here ? This is session detail.

niranjan94 commented on app/api/__init__.py:
```diff
@@ -121,32 +121,33 @@
 api.route(SocialLinkList, 'social_link_list', '/events/<int:id>/social-links', '/events/<identifier>/social-links')
 api.route(SocialLinkDetail, 'social_link_detail',
-          '/social-links/<int:id>', '/events/<int:event_id>/social-links')
+          '/social-links/<int:id>', '/events/<int:event_id>/social-links',
```
Why is `/events/<int:event_id>/social-links` and `/events/<event_identifier>/social-links` here ? This is socialLinkDetail

niranjan94 commented on app/api/__init__.py:
```diff
@@ -156,18 +157,20 @@
           '/event-copyright/<int:id>/relationships/event')
 
 # tax
 api.route(TaxList, 'tax_list', '/events/<int:id>/tax', '/events/<identifier>/tax')
```
This can be removed.

enigmaeth commented:
@niranjan94 I am really sorry, but you must please see this PR #3837. These endpoints were already wrongly implemented when I changed  them to event_identifier. So when I came across these, I made this PR. 

See diff here: https://github.com/fossasia/open-event-orga-server/pull/3837/files

It addresses all of your comments. Plus for the tax one, I have made Detail GET only and List POST only, so your last comment can be let go, as we are using both the endpoints effectively. 
I did the same for speakers-calls and event-copyright so that we can use list and detail separately but with the same endpoints. Here's the commit : https://github.com/enigmaeth/open-event-orga-server/commit/de4d15d0556d5bca05f687c0e2f0e08b76927ab9 
@SaptakS can help with explaining this.

I must admit though that after I made PR #3837, I missed updating those changes in this PR( though they will be resolved once this and #3837 get merged :sweat_smile: )!!

Please review #3837 also. Thanks :smile:

niranjan94 commented:
@enigmaeth can you please add the commits of #3837 into this PR itself and close that PR ? Reference that issue here too. These two are closely related and it would be better to review them together. 

Keep two separate commits.

enigmaeth commented:
@niranjan94 Added commit from PR #3837 here. Please check now.

niranjan94 commented:
Thanks @enigmaeth . Apologies for the mis-understanding from my side.

niranjan94 commented on docs/api/api_blueprint.apib:
```diff
@@ -540,8 +540,8 @@ Get a list of events.
                 "can-pay-by-cheque": false,
                 "bank-details": null,
                 "cheque-details": null,
-                "is-sessions-speakers-enabled": false,
```
This is the correct one.

enigmaeth replied:
@niranjan94  Resolved.

niranjan94 commented on docs/api/api_blueprint.apib:
```diff
@@ -794,8 +794,8 @@ Get a single event.
               "can-pay-by-cheque": false,
               "bank-details": null,
               "cheque-details": null,
-              "is-sessions-speakers-enabled": false,
```
This is the correct one.

niranjan94 commented on docs/api/api_blueprint.apib:
```diff
@@ -936,8 +936,8 @@ All other fields are optional. Add attributes you want to modify.
               "can-pay-by-cheque": false,
               "bank-details": null,
               "cheque-details": null,
-              "is-sessions-speakers-enabled": false,
```
This is the correct one.

enigmaeth commented on app/api/social_links.py:
```diff
@@ -59,10 +59,20 @@ def before_create_object(self, data, view_kwargs):
     data_layer = {'session': db.session,
-                  'model': SocialLink}
+                  'model': SocialLink,
+                  'before_get_object': before_get_object
```
Looks like I deleted methods while rebasing. Will update asap now! @niranjan94

enigmaeth commented:
@niranjan94 Could you please review this now?

enigmaeth commented:
@SaptakS @shubham-padia @niranjan94 @poush Could you review this now please? :sweat_smile:

shubham-padia commented on app/api/tax.py:
```diff
@@ -66,10 +69,40 @@ def before_create_object(self, data, view_kwargs):
+    """
+     tax detail by id
+    """
+    def query(self, view_kwargs):
```
please check this. I have not seen query used in ResourceDetail before this.

enigmaeth replied:
@shubham-padia Resolved.

enigmaeth commented:
@niranjan94 Can you please take a look now? Thanks.

shubham-padia commented on app/api/tax.py:
```diff
@@ -66,10 +69,33 @@ def before_create_object(self, data, view_kwargs):
+     tax detail by id
+    """
+    def before_get_object(self, view_kwargs):
+        if view_kwargs.get('event_identifier'):
```
event.id should not be view_kwargs['id'].Also handle this in same code pattern as done below in Line 80 for `event_id`

enigmaeth replied:
@shubham-padia Resolved. It was already implemented like that at several other places. Changed all those endpoints affeted by this PR. More changes are to made when all `id` is replaced by `event_id`. Please check now.

enigmaeth replied:
@shubham-padia I opened an issue #3856 to implement similar blocks for list methods. Check Issue description in #3856 .

enigmaeth commented on app/api/events.py:
```diff
@@ -135,7 +135,7 @@ def validate_timezone(self, data):
                                 self_view='v1.event_social_link',
                                 self_view_kwargs={'id': '<id>'},
                                 related_view='v1.social_link_detail',
```
This is a wrong relation, as event should not be connected to detail, but list. Fixing this.

poush commented on app/api/tax.py:
```diff
@@ -43,19 +45,20 @@ class TaxList(ResourceList):
-        if view_kwargs.get('id'):
-            event = self.session.query(Event).filter_by(id=view_kwargs['id']).one()
+        if view_kwargs.get('event_id'):
+            event = self.session.query(Event).filter_by(id=view_kwargs['event_id']).one()
```
I think it is good to add `ObjectNotFound` check here. That's true that I won't be able to store the record since ForeignKey will not be matched for `event_id` and even before that `.one()` will throw error but I think it will be 500 error instead its should be  404 or unprocessable

enigmaeth replied:
@poush Yes it should be added at other places too. Already opened an issue #3856. 
I have added for some APIs [here](https://github.com/enigmaeth/open-event-orga-server/commits/exceptn), will make PR when done with all.

poush replied:
@enigmaeth Most of the endpoints will be fixed in the process of solving https://github.com/fossasia/open-event-orga-server/issues/3802

enigmaeth replied:
@poush Resolved.
